02-algorithms/05-complexity-determining-complexity/determining_complexity_answers.md

```diff
@@ -0,0 +1,208 @@
+### 1. What's the Big-O of the following algorithm? Submit your work and reasoning with your solution.
+
+```
+# goodbye_world.rb
+def goodbye_world(n)
+ puts "Goodbye World! #{n}"
+end
+```
+
+This algorithm is `Big-O(1)`, because it takes a constant amount of time, no matter
+what value of `n` is used.
+
+### 2. What's the Big-O of the following algorithm? Submit your work and reasoning with your solution.
+
+```
+# find_largest.rb
+def find_largest(collection)
+ largest = collection[0]
+ collection.length.times do |i|
+   if collection[i] >= largest
+     largest = collection[i]
+   end
+ end
+ largest
+end
+```
+
+This algorithm is `Big-O(n)`, because the program cycles through the main loop
+one time for each value from `0` to `n`, where `n` is the size of the collection.
+
+### 3. What's the Big-O of the following algorithm? Submit your work and reasoning with your solution.
+
+```
+# find_largest_2D_array.rb
+def find_largest(collection)
+ largest = collection[0][0]
+ collection.length.times do |i|
+   subcollection = collection[i]
+   subcollection.length.times do |i|
+     if subcollection[i] >= largest
+       largest = subcollection[i]
+     end
+   end
+ end
+ largest
+end
+
+```
+
+This algorithm is again `Big-O(n)`, where `n` is the number of elements in the 2-D
+array. The reason is that as above, it must cycle through every element in the array.
+(Note: if the array has dimensions `m x k`, then `n = m * k` is the total number of
+elements in the array.)
```

Comment:
Good catch. Many students respond with `n^2` basing only on the nested loop without actually looking at what the loops are doing.